cumulusci/utils/http/multi_request.py

```diff
@@ -0,0 +1,121 @@
+from itertools import chain
+from concurrent.futures import as_completed
+from typing import Iterable, Dict
+
+from requests_futures.sessions import FuturesSession
+
+
+class ParallelHTTP:
+    """A parallelized HTTP client as a context manager"""
+
+    def __init__(self, base_url, max_workers=32):
+        self.base_url = base_url
+        self.max_workers = max_workers
+
+    def __enter__(self, *args):
+        self.session = FuturesSession(max_workers=self.max_workers)
+        return self
+
+    def __exit__(self, *args):
+        self.session.close()
+
+    def _async_request(self, path, method, json=None, headers={}):
```

Comment:
@prescod - If we mutate the default value for `headers`, wouldn't we also be mutating it for all future calls to the function? I'm remember reading something about assigning a mutable default value in function args can be weird.

Reply:
On the one hand, mutable and mutated function default function parameters are indeed very dangerous in Python.

On the other hand, I don't mutate in this case, I just bind a new variable with the same name and copy to it.

On the third hand, it's best to avoid mutable parameters in case some future maintainer (perhaps me!) is not careful or "simplifies" the code by mutating rather than copying.

So no, the code isn't buggy but yes I'll change it to ensure it stays that way.

Reply:
9bb5cbe77615a4c40fe51741ca188a317b57ab87